src/agent/communicator/src/communicator.cpp

```diff
@@ -0,0 +1,159 @@
+        http::response<http::dynamic_body> res = sendHttpRequest(http::verb::post, "/agents", m_token, bodyJson.dump());
+        return res.result_int();
+    }
+
+    http::response<http::dynamic_body> Communicator::sendHttpRequest(http::verb method,
+                                                                     const std::string& url,
+                                                                     const std::string& token,
+                                                                     const std::string& body)
+    {
+        http::response<http::dynamic_body> res;
+        try
+        {
+            boost::asio::io_context io_context;
+            tcp::resolver resolver(io_context);
+            auto const results = resolver.resolve(m_managerIp, m_port);
+
+            tcp::socket socket(io_context);
+            boost::asio::connect(socket, results.begin(), results.end());
+
+            http::request<http::string_body> req {method, url, 11};
+            req.set(http::field::host, "localhost");
+            req.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+            req.set(http::field::accept, "application/json");
+
+            if (!token.empty())
+            {
+                req.set(http::field::authorization, "Bearer " + token);
+            }
+
+            if (!body.empty())
+            {
+                req.set(http::field::content_type, "application/json");
+                req.body() = body;
+                req.prepare_payload();
+            }
+
+            http::write(socket, req);
+
+            beast::flat_buffer buffer;
+
+            http::read(socket, buffer, res);
+
+            std::cout << "Response code: " << res.result_int() << std::endl;
+            std::cout << "Response body: " << beast::buffers_to_string(res.body().data()) << std::endl;
+        }
+        catch (std::exception const& e)
+        {
+            std::cerr << "Error: " << e.what() << std::endl;
+        }
+        return res;
+    }
+
+    const long Communicator::GetTokenRemainingSecs() const
+    {
+        auto now = std::chrono::system_clock::now();
+        auto now_seconds = std::chrono::duration_cast<std::chrono::seconds>(now.time_since_epoch()).count();
+
+        return static_cast<long>(m_tokenExpTimeInSeconds - now_seconds);
+    }
+
+    void Communicator::stop()
+    {
+        {
+            std::lock_guard<std::mutex> lock(m_exitMtx);
+            m_exitFlag = true;
+        }
+
+        m_exitCondVar.notify_one();
+    }
+
+
+    boost::asio::awaitable<void> Communicator::WaitForTokenExpirationAndAuthenticate()
+    {
+        using namespace std::chrono_literals;
+        auto executor = co_await boost::asio::this_coro::executor;
+        boost::asio::steady_timer timer(executor);
+
+        while (true)
+        {
+            SendAuthenticationRequest();
+            auto duration = std::chrono::milliseconds((GetTokenRemainingSecs() - 2)*1000);
```

Comment:
Use a define for the 2 so it is easier to modify it in the future.

Reply:
Done